truss-chains/examples/transcribe/transcribe.py

```diff
@@ -0,0 +1,261 @@
+import asyncio
+import hashlib
+import hmac
+import logging
+import time
+
+import bcp47
+import data_types
+import helpers
+import httpx
+import numpy as np
+import tenacity
+import truss_chains as chains
+
+_IMAGE = docker_image = chains.DockerImage(
+    apt_requirements=["ffmpeg"], pip_requirements=["pandas", "bcp47"]
+)
+
+_WHISPER_URL = "https://model-5woz91z3.api.baseten.co/production/predict"
+_WEBHOOK_URL = "https://model-4q99pkxq.api.baseten.co/development/predict"
+
+
+class DeployedWhisper(chains.StubBase):
+    """Transcribes b64_encoded wave snippets.
+
+    Treat the whisper model like an external third party tool."""
+
+    async def run(self, audio_b64: str) -> data_types.WhisperOutput:
+        resp = await self._remote.predict_async(json_payload={"audio": audio_b64})
+        # TODO: Get Whisper model with langauge, ideally also timestamps.
+        language = "dummy language"
+        bcp47_key = bcp47.languages.get(language.capitalize(), "default")
+        return data_types.WhisperOutput(
+            text=resp["text"], language=language, bcp47_key=bcp47_key
+        )
+
+
+class MacroChunkWorker(chains.ChainletBase):
+    """Downloads and transcribes larger chunks of the total file (~300s)."""
+
+    remote_config = chains.RemoteConfig(
+        docker_image=_IMAGE, compute=chains.Compute(cpu_count=16, memory="32G")
+    )
+    _whisper: DeployedWhisper
+
+    def __init__(
+        self,
+        context: chains.DeploymentContext = chains.provide_context(),
+    ) -> None:
+        super().__init__(context)
+        self._whisper = DeployedWhisper.from_url(
+            _WHISPER_URL,
+            context,
+            options=chains.RPCOptions(retries=3),
+        )
+
+    async def run(
+        self,
+        media_url: str,
+        params: data_types.TranscribeParams,
+        macro_chunk_index: int,
+        start_time: str,
+        duration_sec: int,
+    ) -> data_types.TranscribeOutput:
+        t0 = time.time()
+        tasks = []
+        seg_infos = []
+        logging.debug(f"Macro-chunk [{macro_chunk_index:03}]: Starting.")
+        async with helpers.DownloadSubprocess(
```

Comment:
one note on performance here -- right now, in truss, ability to process items is restricted by `predict_concurrency`, where by default we only allow one request at a time to limit pressure on GPUs. Right now this function will be limited by that, and the machine will be idle while files are downloading (if I'm not misunderstanding this code).

Right now, we solve this with the `preprocess` & `postprocess` functions in truss, which run outside predict lock. We don't have to solve this for this PR, but does it make sense to have pre & post process functions here? Or do you think in chains users should just do their own locking if it's needed

Reply:
We talked about that gotcha when you were here in SF. I resorted to setting `predict_concurrency = num_cpus`.

Reply:
We talked about this gotcha when you visited SF. I resorted to setting `predict_concurrency = num_cpu` and that makes it work well.

Reply:
oh, sorry I forgot. let's discuss this separately, I think this could potentially be an issue for models running on machines with 1 GPU & many CPUs

Reply:
Yeah, it's been already bookmarked as a TODO: https://github.com/basetenlabs/truss/blob/main/truss-chains/truss_chains/code_gen.py#L524